src/lib.rs

```diff
@@ -14,36 +14,43 @@ pub use no_std_net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr, SocketAddrV4, Socke
 /// which knows how to send AT commands to an ESP8266 WiFi module. You could have another implementation
 /// which knows how to driver the Rust Standard Library's `std::net` module. Given this trait, you can how
 /// write a portable HTTP client which can work with either implementation.
-pub trait TcpStack {
+pub trait TcpClient {
 	/// The type returned when we create a new TCP socket
 	type TcpSocket;
 	/// The type returned when we have an error
 	type Error: core::fmt::Debug;
 
-	/// Open a new TCP socket. The socket starts in the unconnected state.
-	fn open(&self) -> Result<Self::TcpSocket, Self::Error>;
+	/// Open a socket for usage as a TCP client.
+	///
+	/// The socket must be connected before it can be used.
+	///
+	/// Returns `Ok(socket)` if the socket was successfully created.
+	fn socket(&self) -> Result<Self::TcpSocket, Self::Error>;
 
 	/// Connect to the given remote host and port.
+	///
+	/// Returns `Ok` if the connection was successful. Otherwise, if the connection could not be
+	/// completed immediately, this function should return [`nb::Error::WouldBlock`].
 	fn connect(
 		&self,
-		socket: Self::TcpSocket,
+		socket: &mut Self::TcpSocket,
 		remote: SocketAddr,
-	) -> Result<Self::TcpSocket, Self::Error>;
+	) -> nb::Result<(), Self::Error>;
 
 	/// Check if this socket is connected
 	fn is_connected(&self, socket: &Self::TcpSocket) -> Result<bool, Self::Error>;
```

Comment:
Should this function stay even with the new `nb::Result` return of `connect`?

Reply:
I would imagine that checking if a socket is currently connected does not require any delay (it would typically be a status register check, software boolean state check, or single register read). As such, I don't think it warrants an `nb::Result` since the check wouldn't be a long operation - open to other use cases here that would warrant an `nb::Result` though!

Reply:
My question was worded confusingly.  I mean, do we need this function at all?  Is there a reason to check if a socket is connected?

Reply:
If the `socket()` call only allocates the required resources, then `connect()` is definitely necessary to check whether the handshake is still in progress/done or if it failed (e.g. timeout).

Reply:
It's also possible for the remote server to close a connection during communication, so we need to be able to indicate that through the stack as well. Typically, the user should not be `send()`ing or `receive()`ing on the socket if it is not connected, as that would generate an error.